File: src/generate/exporter.ts
import { writeFile, readFile } from 'fs/promises';
import { existsSync } from 'fs';
import * as path from 'path';
import Papa from 'papaparse';
import * as yaml from 'js-yaml';
import chalk from 'chalk';
import type { ValidatedCard } from '../types.js';

type CardData = Record<string, string>;

/**
 * Checks if the headers of new data match existing headers.
 * @param existingHeaders - Array of headers from the existing file.
 * @param newHeaders - Array of headers from the new data.
 * @returns true if headers match (ignoring order).
 */
function headersMatch(
  existingHeaders: string[],
  newHeaders: string[],
): boolean {
  if (existingHeaders.length !== newHeaders.length) {
    return false;
  }
  const existingSet = new Set(existingHeaders);
  for (const header of newHeaders) {
    if (!existingSet.has(header)) {
      return false;
    }
  }
  return true;
}

/**
 * Exports selected cards to a file (YAML or CSV) in a format compatible with
 * the anki-llm import command. If the file exists, appends to it after
 * validating schema compatibility.
 *
 * The output is a simple array of objects where keys are Anki field names,
 * matching the format expected by the import command.
 *
 * @param cards - The validated and selected cards to export
 * @param outputPath - The path to the output file
 */
export async function exportCards(
  cards: ValidatedCard[],
  outputPath: string,
): Promise<void> {
  const dataToExport: CardData[] = cards.map((card) => card.ankiFields);
  const ext = path.extname(outputPath).toLowerCase();
  const newCardHeaders =
    dataToExport.length > 0 ? Object.keys(dataToExport[0]) : [];

  let fileContent: string;
  let isAppending = false;

  // Check if file exists and append to it
  if (existsSync(outputPath)) {
    isAppending = true;
    console.log(
      chalk.cyan(
        `\n📦 Appending ${cards.length} card(s) to existing ${outputPath}...`,
      ),
    );

    const existingContent = await readFile(outputPath, 'utf-8');
    let existingData: CardData[] = [];

    // Read and validate existing file
    try {
      if (ext === '.yaml' || ext === '.yml') {
        const parsed = yaml.load(existingContent);
        if (Array.isArray(parsed)) {
          existingData = parsed as CardData[];
        } else if (parsed) {
          throw new Error('Existing YAML is not an array.');
        }
      } else if (ext === '.csv') {
        if (existingContent.trim()) {
          const parsed = Papa.parse<CardData>(existingContent, {
            header: true,
            skipEmptyLines: true,
          });
          if (parsed.errors.length > 0) {
            throw new Error(
              `Failed to parse existing CSV: ${parsed.errors[0].message}`,
            );
          }
          existingData = parsed.data;
        }
      } else {
        throw new Error(
          `Unsupported output format: ${ext}. Please use .csv, .yaml, or .yml.`,
        );
      }
    } catch (e) {
      throw new Error(
        `Failed to parse existing file at ${outputPath}. It may be corrupted or in the wrong format. Error: ${(e as Error).message}`,
      );
    }

    // Schema validation
    if (existingData.length > 0) {
      const existingHeaders = Object.keys(existingData[0]);
      if (!headersMatch(existingHeaders, newCardHeaders)) {
        throw new Error(
          `Schema mismatch: Cannot append cards.\n` +
            `Existing file fields: ${existingHeaders.join(', ')}\n` +
            `New card fields:      ${newCardHeaders.join(', ')}`,
        );
      }
    }

    // Combine data
    const combinedData = [...existingData, ...dataToExport];
    if (ext === '.yaml' || ext === '.yml') {
      fileContent = yaml.dump(combinedData);
    } else {
      // .csv
      fileContent = Papa.unparse(combinedData);
    }

    console.log(
      chalk.gray(
        `  Found ${existingData.length} existing card(s), adding ${cards.length} new card(s)`,
      ),
    );
  } else {
    // File doesn't exist, create new
    console.log(
      chalk.cyan(`\n📦 Exporting ${cards.length} card(s) to ${outputPath}...`),
    );

    if (ext === '.yaml' || ext === '.yml') {
      fileContent = yaml.dump(dataToExport);
    } else if (ext === '.csv') {
      fileContent = Papa.unparse(dataToExport);
    } else {
      throw new Error(
        `Unsupported output format: ${ext}. Please use .csv, .yaml, or .yml.`,
      );
    }
  }

  await writeFile(outputPath, fileContent, 'utf-8');

  console.log(
    chalk.green(
      `\n✓ Successfully ${isAppending ? 'appended to' : 'exported cards to'} ${outputPath}`,
    ),
  );
  console.log(chalk.gray('\nTo import this file into Anki, run:'));
  console.log(
    chalk.white(`  anki-llm import "${outputPath}" --deck "Your Deck Name"`),
  );
}

File: src/generate/exporter.rs
use std::path::Path;

use indexmap::IndexMap;

use super::cards::ValidatedCard;

/// Export cards to a file. Appends if the file already exists (with schema
/// validation). Supports .yaml, .yml, and .csv extensions.
pub fn export_cards(cards: &[ValidatedCard], output_path: &Path) -> Result<(), anyhow::Error> {
    let ext = output_path
        .extension()
        .and_then(|e| e.to_str())
        .unwrap_or("")
        .to_lowercase();

    let data: Vec<IndexMap<String, String>> = cards.iter().map(|c| c.anki_fields.clone()).collect();

    if data.is_empty() {
        anyhow::bail!("No cards to export");
    }

    let new_headers: Vec<&str> = data[0].keys().map(|k| k.as_str()).collect();

    // Load existing data if file exists
    let mut existing: Vec<IndexMap<String, String>> = Vec::new();
    if output_path.exists() {
        let content = std::fs::read_to_string(output_path)?;
        existing = match ext.as_str() {
            "yaml" | "yml" => serde_yaml::from_str(&content)?,
            "csv" => {
                let mut reader = csv::Reader::from_reader(content.as_bytes());
                let mut rows = Vec::new();
                for result in reader.deserialize() {
                    rows.push(result?);
                }
                rows
            }
            _ => anyhow::bail!("Unsupported format: .{ext}. Use .csv, .yaml, or .yml"),
        };

        // Validate headers match
        if let Some(first) = existing.first() {
            let existing_headers: Vec<&str> = first.keys().map(|k| k.as_str()).collect();
            let existing_set: std::collections::HashSet<&str> =
                existing_headers.iter().copied().collect();
            let new_set: std::collections::HashSet<&str> = new_headers.iter().copied().collect();
            if existing_set != new_set {
                anyhow::bail!(
                    "Schema mismatch: existing fields [{}], new fields [{}]",
                    existing_headers.join(", "),
                    new_headers.join(", ")
                );
            }
        }

        eprintln!(
            "\nAppending {} card(s) to existing {} ({} existing)...",
            cards.len(),
            output_path.display(),
            existing.len()
        );
    } else {
        eprintln!(
            "\nExporting {} card(s) to {}...",
            cards.len(),
            output_path.display()
        );
    }

    let combined: Vec<IndexMap<String, String>> = existing.into_iter().chain(data).collect();

    let content = match ext.as_str() {
        "yaml" | "yml" => serde_yaml::to_string(&combined)?,
        "csv" => {
            let mut wtr = csv::Writer::from_writer(Vec::new());
            for row in &combined {
                wtr.serialize(row)?;
            }
            String::from_utf8(wtr.into_inner()?)?
        }
        _ => anyhow::bail!("Unsupported format: .{ext}. Use .csv, .yaml, or .yml"),
    };

    std::fs::write(output_path, content)?;

    eprintln!(
        "\nSuccessfully {} {}",
        if output_path.exists() {
            "appended to"
        } else {
            "exported cards to"
        },
        output_path.display()
    );
    eprintln!("\nTo import into Anki, run:");
    eprintln!(
        "  anki-llm import \"{}\" --deck \"Your Deck Name\"",
        output_path.display()
    );

    Ok(())
}
